src/Components/Server/test/Circuits/RemoteRendererTest.cs

```diff
@@ -419,6 +426,239 @@ public async Task ThrowsIfWeReceivedAnAcknowledgeForANeverProducedBatch()
+        {
+            Type = RootComponentOperationType.Update,
+            ComponentId = component1Id,
+            Marker = CreateMarker(typeof(TestComponent) /* Note the incorrect component type */, new()
+            {
+                [nameof(DynamicallyAddedComponent.Message)] = "Updated message",
+            }),
+        };
+        var operationsJson = JsonSerializer.Serialize(
+            new[] { operation },
+            ServerComponentSerializationSettings.JsonSerializationOptions);
+
+        // Act
+        await renderer.Dispatcher.InvokeAsync(() => renderer.UpdateRootComponents(operationsJson));
+
+        // Assert
+        Assert.Equal(expectedMessage, component1.Message);
+    }
+
+    [Fact]
+    public async Task UpdateRootComponents_Throws_WhenUpdatingComponentFromInvalidDescriptor()
+    {
+        // Arrange
+        var serviceProvider = CreateServiceProvider();
+        var renderer = GetRemoteRenderer(serviceProvider);
+        var component = new DynamicallyAddedComponent()
+        {
+            Message = "Existing message",
+        };
+        var componentId = renderer.AssignRootComponentId(component);
+        var operation = new RootComponentOperation<ServerComponentMarker>
+        {
+            Type = RootComponentOperationType.Update,
+            ComponentId = componentId,
+            Marker = new()
+            {
+                Descriptor = "some random text",
+            },
+        };
+        var operationsJson = JsonSerializer.Serialize(
+            new[] { operation },
+            ServerComponentSerializationSettings.JsonSerializationOptions);
+
+        // Act
+        var task = renderer.Dispatcher.InvokeAsync(() => renderer.UpdateRootComponents(operationsJson));
+
+        // Assert
+        var ex = await Assert.ThrowsAsync<InvalidOperationException>(async () => await task);
+        Assert.StartsWith("Failed to deserialize a component descriptor when updating", ex.Message);
+    }
+
+    [Fact]
+    public async Task UpdateRootComponents_CanRemoveExistingRootComponent()
+    {
+        // Arrange
+        var serviceProvider = CreateServiceProvider();
+        var renderer = GetRemoteRenderer(serviceProvider);
+        var component = new DynamicallyAddedComponent();
+        var componentId = renderer.AssignRootComponentId(component);
+        var operation = new RootComponentOperation<ServerComponentMarker>
+        {
+            Type = RootComponentOperationType.Remove,
+            ComponentId = componentId,
+        };
+        var operationsJson = JsonSerializer.Serialize(
+            new[] { operation },
+            ServerComponentSerializationSettings.JsonSerializationOptions);
+
+        // Act
+        await renderer.Dispatcher.InvokeAsync(() => renderer.UpdateRootComponents(operationsJson));
+
+        // Assert
+        await component.WaitForDisposeAsync().WaitAsync(Timeout); // Will timeout and throw if not disposed
+    }
```

Comment:
Would it also work to add a test here showing the server won't let you add *too many* root components without disposing old ones? Not sure if that enforcement happens within the relevant layer for this unit test though.